1. Загрузите выборку Wine по адресу https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data
2. Извлеките из данных признаки и классы. Класс записан в первом
столбце (три варианта), признаки — в столбцах со второго по по-
следний. Более подробно о сути признаков можно прочитать по ад-
ресу https://archive.ics.uci.edu/ml/datasets/Wine (см. также файл
wine.names, приложенный к заданию)

In [23]:
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import scale

data = pandas.read_csv('wine.data', names=['class', 'Alchohol', 'Malic acid', 'Ash', 'Alcalinity of ash', 'Magnesium', 
                                          'Total phenols', 'Flavanoids', 'Nonflavanoid phenols', 'Proanthocyanins', 'Color intensity', 'Hue',
                                          'OD280/OD315 of diluted wines', 'Proline'])
print(data)

     class  Alchohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0        1     14.23        1.71  2.43               15.6        127   
1        1     13.20        1.78  2.14               11.2        100   
2        1     13.16        2.36  2.67               18.6        101   
3        1     14.37        1.95  2.50               16.8        113   
4        1     13.24        2.59  2.87               21.0        118   
..     ...       ...         ...   ...                ...        ...   
173      3     13.71        5.65  2.45               20.5         95   
174      3     13.40        3.91  2.48               23.0        102   
175      3     13.27        4.28  2.26               20.0        120   
176      3     13.17        2.59  2.37               20.0        120   
177      3     14.13        4.10  2.74               24.5         96   

     Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0             2.80        3.06                  0.28          

In [16]:
classes = data['class']
signs = data.iloc[:, 1:]

3. Оценку качества необходимо провести методом кросс-валидации по
5 блокам (5-fold). Создайте генератор разбиений, который переме-
шивает выборку перед формированием блоков (shuffle=True). Для воспроизводимости результата, создавайте генератор KFold с фик-
сированным параметром random_state=42. В качестве меры каче-
ства используйте долю верных ответов (accuracy).

In [17]:
cv = KFold(n_splits=5, shuffle=True, random_state=42)

4. Найдите точность классификации на кросс-валидации для метода
k ближайших соседей (sklearn.neighbors.KNeighborsClassifier), при
k от 1 до 50. При каком k получилось оптимальное качество? Чему
оно равно (число в интервале от 0 до 1)? Данные результаты и
будут ответами на вопросы 1 и 2.]

In [19]:
best_mean = 0
best_k = 0
for k in range(1, 51):
    model = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(model, signs, classes, cv = cv, scoring='accuracy')
    mean = scores.mean()
    if mean > best_mean:
        best_mean = mean
        best_k = k
print(best_k)
print(best_mean)

1
0.7304761904761905


5. Произведите масштабирование признаков с помощью функции
sklearn.preprocessing.scale. Снова найдите оптимальное k на кросс-
валидации.

In [24]:
signs_scaled = scale(signs)
signs_scaled_DF = pd.DataFrame(signs_scaled, columns=signs.columns)
print(signs_scaled_DF)

     Alchohol  Malic acid       Ash  Alcalinity of ash  Magnesium  \
0    1.518613   -0.562250  0.232053          -1.169593   1.913905   
1    0.246290   -0.499413 -0.827996          -2.490847   0.018145   
2    0.196879    0.021231  1.109334          -0.268738   0.088358   
3    1.691550   -0.346811  0.487926          -0.809251   0.930918   
4    0.295700    0.227694  1.840403           0.451946   1.281985   
..        ...         ...       ...                ...        ...   
173  0.876275    2.974543  0.305159           0.301803  -0.332922   
174  0.493343    1.412609  0.414820           1.052516   0.158572   
175  0.332758    1.744744 -0.389355           0.151661   1.422412   
176  0.209232    0.227694  0.012732           0.151661   1.422412   
177  1.395086    1.583165  1.365208           1.502943  -0.262708   

     Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0         0.808997    1.034819             -0.659563         1.224884   
1         0.568648    0.7

In [25]:
best_mean = 0
best_k = 0
for k in range(1, 51):
    model = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(model, signs_scaled_DF, classes, cv = cv, scoring='accuracy')
    mean = scores.mean()
    if mean > best_mean:
        best_mean = mean
        best_k = k
print(best_k)
print(best_mean)

29
0.9776190476190475


6. Какое значение k получилось оптимальным после приведения при-
знаков к одному масштабу? Как изменилось значение качества?
Приведите ответы на вопросы 3 и 4.

Оптимальное к после масштабирования признаков = 29. Значение качества увеличилось с 0.73 (без масштабирования) до 0.98